ugvc/pipelines/single_cell_qc/collect_statistics.py

```diff
@@ -0,0 +1,268 @@
+    merged_histogram_csv = merge_trimmer_histograms(
+        input_files.trimmer_histogram_csv, output_path=output_path
+    )
+
+    # Read inputs into df
+    trimmer_stats = pd.read_csv(input_files.trimmer_stats_csv)
+    df_trimmer_failure_codes = read_trimmer_failure_codes(
+        input_files.trimmer_failure_codes_csv,
+        add_total=True,
+    )
+    histogram = pd.read_csv(merged_histogram_csv)
+    sorter_stats = read_sorter_statistics_csv(input_files.sorter_stats_csv)
+    star_stats = read_star_stats(input_files.star_stats)
+    star_reads_per_gene = pd.read_csv(input_files.star_reads_per_gene, header=None, sep="\t")
+
+    # Get insert subsample quality and lengths
+    insert_quality, insert_lengths = get_insert_properties(input_files.insert_subsample)
+
+    # Save statistics into h5
+    output_filename = Path(output_path) / OutputFiles.H5.value
+
+    with pd.HDFStore(output_filename, "w") as store:
+        store.put(H5Keys.TRIMMER_STATS.value, trimmer_stats, format="table")
+        store.put(H5Keys.TRIMMER_FAILURE_CODES.value, df_trimmer_failure_codes, format="table")
+        store.put(H5Keys.TRIMMER_HISTOGRAM.value, histogram, format="table")
+        store.put(H5Keys.SORTER_STATS.value, sorter_stats, format="table")
+        store.put(H5Keys.STAR_STATS.value, star_stats, format="table")
+        store.put(H5Keys.STAR_READS_PER_GENE.value, star_reads_per_gene, format="table")
+        store.put(H5Keys.INSERT_QUALITY.value, insert_quality, format="table")
+        store.put(H5Keys.INSERT_LENGTHS.value, pd.Series(insert_lengths), format="table")
+
+    return output_filename
+
+
+def read_star_stats(star_stats_file: str) -> pd.DataFrame:
+    """
+    Read STAR stats file (Log.final.out) and return parsed DataFrame
+
+    :param star_stats_file: path to STAR stats file
+
+    :return: DataFrame with parsed STAR stats
+    """
+    df = pd.read_csv(star_stats_file, header=None, sep="\t")
+    df.columns = ["metric", "value"]
+
+    # parse metric description
+    df["metric"] = df["metric"].str.replace("|", "").str.strip().str.replace(" ", "_")
+    df.loc[:, "metric"] = df["metric"].str.replace(",", "").str.replace(":", "")
+
+    # Add read_type (general, unique_reads, multi_mapping_reads, unmapped_reads, chimeric_reads)
+    df.loc[:, "read_type"] = (
+        df["metric"]
+        .where(df["value"].isnull())
+        .ffill()
+        .fillna("general")
+        .str.lower()
+        .str.replace(":", "")
+    )
+    df = df.dropna(subset=["value"])
+
+    # parse value: add value type (number, percentage, datetime)
+    df["value_type"] = "number"
+    df.loc[df["value"].str.endswith("%"), "value_type"] = "percentage"
+    df.loc[df["value"].str.find(":") != -1, "value_type"] = "datetime"
+    df["value"] = df["value"].str.replace("%", "")
+    return df
+
+
+def get_insert_properties(insert_subsample, max_reads=None) -> Union[pd.DataFrame, list[int]]:
```

Comment:
since python 3.10 the recommended way is to use |
```suggestion
def get_insert_properties(insert_subsample, max_reads=None) -> pd.DataFrame | list[int]:
```